rockchip: add support for Pine64 RockPro64

Manawyrm opened the pull request:
This PR adds the new rockchip target, support for RK3399 SoC and the Pine64 RockPro64 device. 
It also includes the required u-boot and arm-trusted-firmware packages.

Manawyrm committed:
arm-trusted-firmware-rockchip: add new package

This is needed to build the uboot-rockchip, needed for the rockchip target

Signed-off-by: Tobias Mädel <t.maedel@alfeld.de>
Tested-by: Tobias Schramm <t.schramm@manjaro.org>

Manawyrm committed:
uboot-rockchip: add new package

This package is needed for the rockchip target.

Signed-off-by: Tobias Mädel <t.maedel@alfeld.de>
Tested-by: Tobias Schramm <t.schramm@manjaro.org>

Manawyrm committed:
rockchip: add support for Pine64 RockPro64

This adds the new rockchip target and support for RockPro64 RK3399

Flash:    16 MiB SPI NOR
RAM:      2 GiB/4 GiB LPDDR4
SoC:      RK3399
USB:      2x USB 2.0, 1x USB 3.0, 1x USB-C
Ethernet: 1x GbE
PCIe:     PCIe 2.0, 4 lanes
Storage:  eMMC or SD card
Optional SDIO wifi/bt module

The Pine64 RockPro64 is a single-board-computer with a 4x PCIe connector,
6 ARM64 cores (4 little, 2 big), plenty of RAM and storage.

By default the single Gigabit-Ethernet port is configured as the
LAN port.

Installation of the firware is possible by dd'ing the image
to an SD card or the eMMC flash.

Serial: 3v3 1500000 8n1

U-boot is build from the mainline tree and
integrated into the images. Required ATF to build u-boot
is downloaded from a CI build bot.

Signed-off-by: Tobias Mädel <t.maedel@alfeld.de>
Tested-by: Tobias Schramm <t.schramm@manjaro.org>

adrianschmutzler commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,31 @@
+	uboot-envtools \
+	kmod-usb-hid \
+	kmod-fs-vfat kmod-nls-cp437 kmod-nls-iso8859-1 \
+	partx-utils mkf2fs e2fsprogs
```
Either multiple packages per line up to 80 chars or one package per line. I'd prefer the first.

Manawyrm replied:
Fixed, thanks!

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,87 @@
+	BOOT_SCRIPT := sd
+	IMAGE/sysupgrade.img.gz := boot-common | boot-script | pine64-img | gzip | append-metadata
+endef
+ifeq ($(SUBTARGET),rk3399)
```
That's inconsequent. Either put the full device (i.e. the definition) into the subtarget brackets or do not use a subtarget.

If you want to have the subtarget, I'd recommend to use a separate file for its devices, like it's done for sunxi for example: https://github.com/openwrt/openwrt/blob/master/target/linux/sunxi/image/Makefile#L45

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,87 @@
+
+### Devices ###
+define Device/Default
+	DEVICE_VENDOR := Rockchip
```
I don't think the vendor should go into the default node. (Targets should be per-architecture, not per-vendor.)
Please move it to the devices.

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,87 @@
+	IMAGES := sysupgrade.img.gz
+endef
+
+define Device/rockpro64-sd
```
The syntax for devices is vendor_model-variant.

So, I assume you would want to use "rockchip_rockpro64-sd" here.

adrianschmutzler replied:
Hmm, the compatible is pine64,rockpro64, so we should sort out what's the appropriate vendor name first.

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,87 @@
+endef
+
+define Device/rockpro64-sd
+	DEVICE_MODEL := rockpro64-sd
```
Looks like that should be:
DEVICE_MODEL := RockPro64
DEVICE_VARIANT := SD

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,87 @@
+	TARGET_DEVICES += rockpro64-emmc
+endif
+
+$(eval $(call BuildImage))
```
Missing new line at EOF.

adrianschmutzler commented on target/linux/rockchip/image/gen_pine64_img.sh:
```diff
@@ -0,0 +1,50 @@
+# Copy the idbloader and the u-boot image to the image at sector 0x40 and 0x4000
+dd if="$STAGINGDIR"/${DEVICENAME}-idbloader.img of="$OUTPUT" seek=64 conv=notrunc
+dd if="$STAGINGDIR"/${DEVICENAME}-u-boot.itb of="$OUTPUT" seek=16384 conv=notrunc
+
```
Unneeded empty line at EOF.

adrianschmutzler commented on target/linux/rockchip/modules.mk:
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright (C) 2019 OpenWrt.org
```
I'm not really a fan of copying those (outdated) copyright lines over from source files, particularly when it's trivial files.

adrianschmutzler commented on target/linux/rockchip/base-files/lib/preinit/79_move_config:
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
```
I'm not sure whether preinit files are executed or sourced ...

Manawyrm replied:
They are sourced indeed (see /etc/preinit), but every preinit file on my OpenWRT installation has a shebang, so I think I better leave this in?

adrianschmutzler replied:
Well, most people just don't care. But I personally prefer to have this tidy, just like you wouldn't give the file an executable bit without need.

Manawyrm replied:
OK, I've changed it now :)

adrianschmutzler commented on target/linux/rockchip/base-files/lib/preinit/79_move_config:
```diff
@@ -0,0 +1,20 @@
+}
+
+boot_hook_add preinit_mount_root move_config
+
```
Remove newline at EOF.

adrianschmutzler commented on target/linux/rockchip/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
```
If you use a subtarget, please move all base-files to the subtarget directory, so they are not automatically used when a new subtarget is added.

adrianschmutzler commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,29 @@
+endef
+
+include $(INCLUDE_DIR)/target.mk
+DEFAULT_PACKAGES += \
```
Missing empty line before.

adrianschmutzler commented on target/linux/rockchip/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,21 @@
+
+. /lib/functions/uci-defaults.sh
+. /lib/functions.sh
+. /lib/functions/system.sh
```
At the moment, system.sh is not used.

adrianschmutzler commented on target/linux/rockchip/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+# Copyright (C) 2014-2016 OpenWrt.org
+# Copyright (C) 2017 LEDE project
```
I would not add them at all, but if, it should be 2020 Openwrt.org

Manawyrm commented:
Should all be fixed now. 
@adrianschmutzler Could you take a second look at it?

Thanks!

adrianschmutzler commented on target/linux/rockchip/image/rk3399.mk:
```diff
@@ -0,0 +1,28 @@
+	DEVICE_MODEL := RockPro64
+	DEVICE_VARIANT := emmc
+	DEVICE_DTS := rockchip/rk3399-rockpro64
+	SUPPORTED_DEVICES := pine64,rockpro64
```
This will allow crossflashing, but as I understand the PR, this is intended?

Manawyrm replied:
Yes, this is expected, it's the same device, only with different storage media.

ynezz replied:
I'm probably missing something, but to me it seems like an overkill. What about supporting both eMMC/SD with single bootscript instead?

ynezz replied:
See for example 5f2ff607e286ab92ede986d62b39391557d36c61

adrianschmutzler replied:
I'd prefer ynezz solution as well, particularly if this enables us to have one device for one device.

Manawyrm replied:
Same reason as below. Let's have only one issue for this. I'm closing this one, let's talk in the other one :)

ynezz replied:
It seems like this device is using distroboot, so you should be able to find out from which device was your script loaded via devtype and devnum variables.

ynezz replied:
This is how your boot script is probably called `boot_a_script=load ${devtype} ${devnum}:${distro_bootpart} ${scriptaddr} ${prefix}${script}; source ${scriptaddr}`

Manawyrm replied:
Oh, thanks. Very good suggestion, that's much neater than my 2 targets. 
It's fixed now, please take another look!

ynezz replied:
Just a nitpick, you should probably rename `mmc.bootscript` to `bootscript` now as it supports mmc/sd.

Manawyrm replied:
hm, I renamed it from emmc.bootscript to mmc.bootscript because it can now handle both types of mmc devices.

adrianschmutzler commented on target/linux/rockchip/image/rk3399.mk:
```diff
@@ -0,0 +1,28 @@
+	IMAGE/sysupgrade.img.gz := boot-common | boot-script | pine64-img | gzip | append-metadata
+endef
+
+TARGET_DEVICES += pine64_rockpro64-emmc
```
Those should follow their definitions directly, with no newline between and one newline after. Just have a look at other targets if you are unsure.

Manawyrm replied:
Fixed, thanks!

lucize commented:
@Manawyrm 
hi, I'm trying this on ubuntu 18.04 and I get 
`make[5]: arm-none-eabi-gcc: Command not found` do I miss something ?

Manawyrm commented:
@lucize 
> `make[5]: arm-none-eabi-gcc: Command not found` do I miss something ?

Hi,
both the @TobleMiner and me had arm-none-eabi-gcc toolchains available on our systems while developing and testing this, so it might be possible that there is something missing here to use OpenWRTs host toolchain. I don't really believe that to be the case, because the uboot and atf packages are both pretty much copies of existing sunxi packages, but you never know... 

Can you tell me which commands you ran exactly and post the output (preferably the whole build log) as a gist? 

Thanks for testing!

lucize commented:
make V=s -j12 and after that -j1
here are last lines, do you really need all the build log ?
[last_lines.txt](https://github.com/openwrt/openwrt/files/4471411/last_lines.txt)

Manawyrm commented:
@lucize 
> [last_lines.txt](https://github.com/openwrt/openwrt/files/4471411/last_lines.txt)

OK, thanks, that helps. So the ATF is really missing it's toolchain to build. 
I'm not sure why this isn't happening for arm-trusted-firmware-sunxi and arm-trusted-firmware-mvebu as well. 

I will take a look at it and push a patch.

lucize commented:
```
 sudo updatedb
[sudo] password for build:
build@build:~/source$ locate arm-none-eabi-gcc
build@build:~/source$

```

Manawyrm commented:
@lucize If you just want to test the image itself, you can install
Debian: https://packages.debian.org/de/buster/gcc-arm-none-eabi
Arch: https://www.archlinux.org/packages/community/x86_64/arm-none-eabi-gcc/

But I think this should work without external dependencies. I'll make sure to fix this shortly.

diizzyy commented on package/boot/arm-trusted-firmware-rockchip/Makefile:
```diff
@@ -0,0 +1,51 @@
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL=https://github.com/ARM-software/arm-trusted-firmware
+PKG_SOURCE_DATE:=2020-01-03
```
Why are we not using releases here which are much easier to track?

Manawyrm replied:
I need commit https://github.com/ARM-software/arm-trusted-firmware/commit/b4899041e5f0b8e8b388c6511b5233516b8785ec which isn't part of any stable release yet. 
I could use 2.3.0-rc0 if that would be preferrable?

diizzyy replied:
Yes, that's a much better choice

Manawyrm replied:
Done.

diizzyy commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -0,0 +1,54 @@
+#
+# Copyright (C) 2013-2016 OpenWrt.org
```
Remove?
https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md#makefile-contents-should-contain

Manawyrm replied:
OK, I've removed the copyright headers.

diizzyy commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (C) 2012-2020 OpenWrt.org
+# Copyright (C) 2017 LEDE project
```
Where does this come from?
Remove?

diizzyy commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,61 @@
+# 
+# Copyright (C) 2020 Tobias Maedel
+# Copyright (C) 2012-2020 OpenWrt.org
+# Copyright (C) 2016-2017 LEDE project
```
Same as earlier...

diizzyy commented on target/linux/rockchip/image/gen_pine64_img.sh:
```diff
@@ -0,0 +1,49 @@
+#!/bin/sh
+# 
+# Copyright (C) 2020 Tobias Maedel
+# Copyright (C) 2012-2020 OpenWrt.org
```
Same as earlier

diizzyy commented on target/linux/rockchip/rk3399/target.mk:
```diff
@@ -0,0 +1,13 @@
+#
+# Copyright (C) 2020 OpenWrt.org
```
Same as above

diizzyy commented on target/linux/rockchip/rk3399/target.mk:
```diff
@@ -0,0 +1,13 @@
+ARCH:=aarch64
+SUBTARGET:=rk3399
+BOARDNAME:=RK3399 boards (64 bit)
+CPU_TYPE:=cortex-a72
```
This should most likely be set to ```cortex-a53``` due to the little cluster

Manawyrm replied:
OK! Done.

diizzyy commented:
Unless I'm reading the image generation part wrong ```IMAGE/sysupgrade.img.gz``` is a bit wierd and inconsistent naming.

diizzyy commented on package/boot/arm-trusted-firmware-rockchip/Makefile:
```diff
@@ -0,0 +1,51 @@
+    DEPENDS:=@TARGET_rockchip_rk3399
+endef
+
+export GCC_HONOUR_COPTS=s
```
A nicer way would be to set and/or strip the optimization flag?

Manawyrm replied:
I'm not sure if I can influence this in the external build system of the ATF. 
OpenWRT does this in the include for u-boot as well: 
https://github.com/openwrt/openwrt/blob/master/include/u-boot.mk

arm-trusted-firmware-sunxi does so as well: 
https://github.com/openwrt/openwrt/blob/master/package/boot/arm-trusted-firmware-sunxi/Makefile#L34

If you know a nicer way on how to fix this, please let me know.

ynezz replied:
This looks like Linux only solution, see our remarks in https://github.com/openwrt/openwrt/pull/2521

ynezz replied:
This looks like Linux only solution, see our remarks in https://github.com/openwrt/openwrt/pull/2521

Manawyrm replied:
See https://github.com/openwrt/openwrt/pull/2917#issuecomment-614329227

Manawyrm commented:
> Unless I'm reading the image generation part wrong `IMAGE/sysupgrade.img.gz` is a bit wierd and inconsistent naming.

What exactly do you mean by that? Many other targets (sunxi, bcm27xx, etc) use the same value. 
The resulting images in the bin/targets/ dir looks like this: 
`openwrt-rockchip-rk3399-pine64_rockpro64-emmc-ext4-sysupgrade.img.gz`

I could rename it to sdcard.img.gz and emmc.img.gz?

diizzyy commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,30 @@
+include $(INCLUDE_DIR)/target.mk
+
+DEFAULT_PACKAGES += \
+	uboot-envtools kmod-usb-hid kmod-fs-vfat kmod-nls-cp437 \
```
Is vfat needed as you already have fat support in kernel config?
Same goes for nls packages and usb-hid...

Manawyrm replied:
Yep, good idea.

diizzyy commented:
For consistency you should preferably follow sunxi naming
https://downloads.openwrt.org/snapshots/targets/sunxi/cortexa53/

diizzyy commented on target/linux/rockchip/rk3399/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,18 @@
+#!/bin/sh
+
+. /lib/functions/uci-defaults.sh
+. /lib/functions.sh
```
Use this as a template? Slightly cleaner
https://raw.githubusercontent.com/openwrt/openwrt/master/target/linux/sunxi/base-files/etc/board.d/02_network

Manawyrm replied:
Done.

Manawyrm commented:
@diizzyy 
> For consistency you should preferably follow sunxi naming
> https://downloads.openwrt.org/snapshots/targets/sunxi/cortexa53/

Am I right in thinking that every subtarget only gets one kernel for the complete subtarget? 
Is it preferable to have the same kernel for both RK3328 and RK3399 SoCs, the RK3399 has big cores with a72 instructions, the RK3328 is a pure a53 SoC. 

Does any other target already use big.little CPUs? 
I'm not sure how the different instruction sets are handled on big.little CPUs, so far I'm compiling everything with the a72 target and haven't encountered any issues. Would a72 code crash if the scheuduler decides to execute it on a little core, or is there more clever handling (like moving the process to the big cores when "illegal" instructions are found)?

Manawyrm commented:
The issue reported by @lucize is a bit bigger unfortunately. 
I have updated the arm-trusted-firmware-rockchip package accordingly to deal with this issue.

The ATF consists of 2 parts: 
- the native (aarch64, armv8) code for DDR init and general init
- firmware for an integrated Cortex-M0 32-bit microprocessor

This 32-bit M0 CPU needs it's firmware to be compiled with a arm-none-eabi-gcc, not with the aarch64- toolchain and also not with a -linux toolchain. 

The Marvell mvebu target has this same issue (although their embedded MCU is an M3): 
https://github.com/openwrt/openwrt/blob/master/package/boot/arm-trusted-firmware-mvebu/Makefile
They solved this issue by downloading a arm-none-eabi- toolchain from Linaro. 

I have updated my atf-rockchip package to do the same thing and download a (newer) none-eabi 32-bit toolchain. 

@adrianschmutzler @diizzyy Could you look at the ATF package again and see if this is OK for you?

diizzyy commented on package/boot/arm-trusted-firmware-rockchip/Makefile:
```diff
@@ -0,0 +1,73 @@
+PKG_NAME:=arm-trusted-firmware-rockchip
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
```
It would be nicer(?) if you'd grab the generated tarball by Github
https://github.com/openwrt/packages/blob/master/utils/ttyd/Makefile#L14 (Example)

Manawyrm replied:
Hm, maybe nicer in terms of speed and reliability. 
I have now changed it to GitHubs .tar.gz codeload. URL. 
Unfortunately that added an additional folder rename and a bit of shuffling around, but it is now done.

Manawyrm commented:
@lucize Please try your build again with the current version of this patch.
It should work now.

lucize commented:
@Manawyrm no sure if this error is from a standard build dependency

```
  OBJCOPY lib/efi_loader/helloworld.efi
start=$(aarch64-openwrt-linux-musl-nm u-boot | grep __rel_dyn_start | cut -f 1 -d ' '); end=$(aarch64-openwrt-linux-musl-nm u-boot | grep __rel_dyn_end | cut -f 1 -d ' '); tools/relocate-rela u-boot-nodtb.bin 0x00200000 $start $end
  CAT     u-boot-dtb.bin
  COPY    u-boot.bin
  MKIMAGE u-boot.img
  MKIMAGE u-boot-dtb.img
./"arch/arm/mach-rockchip/make_fit_atf.py" \
arch/arm/dts/rk3399-rockpro64.dtb > u-boot.its
  MKIMAGE u-boot.itb
sh: 1: dtc: not found
./tools/mkimage: Can't open u-boot.itb.tmp: No such file or directory
Makefile:1351: recipe for target 'u-boot.itb' failed
make[4]: *** [u-boot.itb] Error 255
make[4]: Leaving directory '/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rockpro64-rk3399/u-boot-2020.01'
Makefile:50: recipe for target '/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rockpro64-rk3399/u-boot-2020.01/.built' failed
make[3]: *** [/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rockpro64-rk3399/u-boot-2020.01/.built] Error 2
make[3]: Leaving directory '/home/build/source/package/boot/uboot-rockchip'
time: package/boot/uboot-rockchip/rockpro64-rk3399/compile#7.86#4.80#12.24
package/Makefile:111: recipe for target 'package/boot/uboot-rockchip/compile' failed
make[2]: *** [package/boot/uboot-rockchip/compile] Error 2
make[2]: Leaving directory '/home/build/source'

```

lucize commented:
```
build@build:~/source/build_dir/target-aarch64_cortex-a53_musl/linux-rockchip_rk3399/linux-5.4.31/scripts/dtc$ ./dtc -v
Version: DTC 1.5.0-g702c1b6c

```

Manawyrm commented:
@lucize That was a bit tricky to fix, but it works now. Please try again with an updated tree.

lucize commented:
maybe $(SED) is better on Build/Configure and is a very common thing to do in openwrt :)

lucize commented:
I've selected the sd profile and emmc sysupgrade file was also present

sadly I can't test as it seems my board is ROCK64 rk3328 not Pine64, but I guess this can be an example for adding support

Manawyrm commented:
> maybe $(SED) is better on Build/Configure and is a very common thing to do in openwrt :)

Good point, thanks. Fixed. 

> ROCK64 rk3328 
If you want to do some testing, I could send you some pre-build images to try. Should be very easy to do. But I would want to get this PR accepted and then do another one with RK3328 support.

lucize commented:
> I could send you some pre-build images to try

I would like to test some packages on it that are in PR state right now
if you have it you can add an another PR and prioritize your needs. I can build test and run them, I don't know if it helps me in the binary state

Manawyrm commented:
I don't seem to have a RK3328 board to test with, unfortunately. Maybe I can get my hands on one later. 
You could try to build the ATF and uboot-packages by just changing everything to rk3328. Then just change the devicename and DTS name in the main-makefile.
The image building process is identical as far as I can tell.

Manawyrm commented:
@diizzyy 
> For consistency you should preferably follow sunxi naming
> https://downloads.openwrt.org/snapshots/targets/sunxi/cortexa53/

I've looked at this a bit more and I'm pretty certain that a seperate subtarget per SoC would be the way to go: 
- the ATF is specific to the SoC (and is built via the subtarget)
- rk3399 has additional PCIe support and USB-Type-C support code

ynezz commented:
> They solved this issue by downloading a arm-none-eabi- toolchain from Linaro.

Nope, its not solved as you can see in #2521

Manawyrm commented:
@ynezz Do you mean the external toolchain? 
It is absolutely required to build the Cortex-M0 firmware (see https://github.com/openwrt/openwrt/pull/2917#issuecomment-613131885). 

It would be possible to build a new host package and compile our own arm-none-eabi toolchain and use that for -rockchip and -mvebu targets. 

Should this be done in a different PR or would you like to have this included in this PR?

Manawyrm commented:
@ynezz To quote a question from the other issue from you: 
> Anyway, why do you think, that this arm-trusted-firmware-mvebu belongs to the master tree? 

Because the ATF is needed to compile the u-boot and the u-boot is needed in the SD card images. 
Without the ATF the devices cannot init their DDR interface and do not work at all. That's why these packages are in master.

ynezz commented:
> It would be possible to build a new host package and compile our own arm-none-eabi toolchain and use that for -rockchip and -mvebu targets.

Indeed, that is one possible solution. IMHO its very expensive one in terms of computing/build resources (which are scarce), because it takes probably ages to build another toolchain just in order to output binary which is small like peanuts.

What about using prebuild binaries? I mean, having some separate repository/project which would create, build and package binary artifacts usable in projects/distributions like for example OpenWrt. This should be of course done in a reproducible way, so the validity of such binary artifacts could be verified.

In ideal world we would probably have both options, possibility to use prebuild binaries or build everything from sources if requested via some .config option. But I assume, that the other unused option (not enabled by default) would bitrot after some time as it wouldn't be thoroughly tested.

ynezz commented:
> Should this be done in a different PR or would you like to have this included in this PR?

I didn't reviewed the other parts of this PR, I've just seen this Linux only bits in this PR so I've expressed my view on it.

Manawyrm commented:
@ynezz Hm :/ 
I don't really like the idea of downloading binaries when we could build it ourselves.

Where should such a repository with binaries be hosted? private GitHub account?
Or maybe we could create a repo in the OpenWRT account that would contain some build scripts and would use CI to build the ATF? 

There isn't any public source of pre-compiled RK3399 ATFs (as far as I'm aware).

The additional toolchain would indeed be a bit CPU intensive to build, but we already have 2 targets requiring it. Also, we probably only need the bare-minimum toolchain (gcc, linker and some misc utils) because it's for a bare-metal Cortex-M0 processor.

TobleMiner commented:
I don't like the idea of using prebuilt ATF binaries, either. What is the policy regarding such blobs in out package builds?

The number of targets requiring ATF builds will only increase. We should establish a common policy (and possibly source) for such builds.

 @blocktrron, @pepe2k what are your opinions on this matter?

ynezz commented on target/linux/rockchip/image/sd.bootscript:
```diff
@@ -0,0 +1,8 @@
+setenv bootargs "console=ttyS2,1500000 console=tty1 earlycon=uart8250,mmio32,0xff1a0000 root=PARTUUID=@ROOT@-02 rw rootwait"
```
Any reason for hardcoding the partuuid instead of getting it from the boot medium via `part uuid` ?

Manawyrm replied:
because this mechanism (more specificly ${mmc_bootdev} ) is sunxi-specific: https://lists.denx.de/pipermail/u-boot/2018-July/334837.html

I don't like it either, but I don't know about a better solution right now.

ynezz replied:
I'm asking about this `root=PARTUUID=@ROOT@-02`, nothing about `$mmc_bootdev`.

Manawyrm replied:
I need to specify the MMC device when calling part uuid. This needs mmc_bootdev again.
But yes, I can remove that hardcoding.

ynezz commented:
> I don't like the idea of using prebuilt ATF binaries, either.
> I don't really like the idea of downloading binaries when we could build it ourselves.

Could any of you provide argument against using of prebuilt binaries? My argument in favor of prebuilt binaries is avoiding waste of build resources on the buildbots, on a package which seems to have release cadence of [6 months](https://github.com/ARM-software/arm-trusted-firmware/releases) and the resulting binary is somewhere in the range of dozen kilobytes (sunxi/a53 bl31.bin is 25017 bytes).

> I don't really like the idea of downloading binaries when we could build it ourselves.

Whats the difference to build toolchain and binary ourselves _once_, then reuse the binary each consecutive build Vs building toolchain and binary _every_ time?

Manawyrm commented:
@ynezz 
>  in favor of prebuilt binaries
OK.
How would you prefer to organise such a thing? 
Should we open a new repository on GitHub for the resulting binaries, compile them and push them there? If so: Who would create such a repo on the OpenWRT account?

blocktrron commented:
@TobleMiner 

This looks like we have no perfect solution to the problem. As it's incompatible with our toolchain, myself would view the usage of an external x86 Linux toolchain as the more hacky solution to the problem.

Building the ATF binaries by some kind of CI seems to be the most feasible compromise in this case. 

OTOH i have to admit that I'm not all into the OpenWrt infrastructure, so I'm not sure whether or not we are able to implement this on our current setup. If it's to much work, someone might setup something similar using GitHub actions to generate the binary output we can drop into the buildsystem. But these are just on the spot ideas to tackle this.

Are you aware whether or not the build output is reproducible? This might ease possible concerns of of using prebuilt binaries.

adrianschmutzler commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,49 @@
+### Devices ###
+define Device/Default
+	KERNEL := kernel-bin
+	IMAGES := sysupgrade.img.gz
```
Add
`  SUPPORTED_DEVICES := $(subst _,$(comma),$(1))`
here and remove SUPPORTED_DEVICES for the actual device.

Manawyrm replied:
Good idea, added, works!

adrianschmutzler commented on target/linux/rockchip/image/rk3399.mk:
```diff
@@ -0,0 +1,15 @@
+define Device/pine64_rockpro64
+	DEVICE_VENDOR := Pine64
+	DEVICE_MODEL := RockPro64
+	DEVICE_DTS := rockchip/rk3399-rockpro64
```
Like SUPPORTED_DEVICES, this could also be derived from node name to provide more consistency:

In Device/Default:
```
  DEVICE_DTS_DIR := rockchip
  DEVICE_DTS = $$(SOC)-$(lastword $(subst _, ,$(1)))
```

Here just:
```
SOC := rk3399
```

While this looks more complicated in the first place, it has the advantage that DTS names are "derived" from node names and thus are forced to be consistent. If you still have to deviate, you can still overwrite DEVICE_DTS locally.

Manawyrm replied:
setting `DEVICE_DTS_DIR` breaks a lot of different things, I think the name is already in use?
anyway: good idea, done!

adrianschmutzler replied:
Yes, DEVICE_DTS_DIR does interesting things.
It actually replaces DTS_DIR:
https://github.com/openwrt/openwrt/blob/master/include/image.mk#L525

So, my example was wrong, and you innovative solution should be okay if all (most of the) future devices in this target will be rockchip subfolder.

Another solution would be to have `DTS_DIR := $(DTS_DIR)/rockchip` , but then it would be actually impossible to overwrite that per-device.

All just FYI, I'm okay with how you did it.

adrianschmutzler commented on target/linux/rockchip/image/rk3399.mk:
```diff
@@ -0,0 +1,15 @@
+# See /LICENSE for more information.
+#
+define Device/pine64_rockpro64
+	DEVICE_VENDOR := Pine64
```
Indent for device definitions is two spaces by convention.

```
define Device/pine64_rockpro64
  DEVICE_VENDOR := Pine64
  ...
```

Manawyrm replied:
Yep, should now be fixed.

adrianschmutzler commented on target/linux/rockchip/rk3399/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,18 @@
+case "$(board_name)" in
+"pine64,rockpro64")
+	ucidef_set_interface_lan 'eth0'
+	;;
```
If it falls into the default case, remove it. If you already have an overview of devices, judge whether the default case is appropriate beforehand.

Manawyrm replied:
OK :)

adrianschmutzler commented on target/linux/rockchip/rk3399/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,18 @@
+
+board_config_flush
+
+exit 0
```
Missing new line at EOL.

Manawyrm replied:
Oops, fixed.

lucize commented:
@Manawyrm I'm trying to build for rk3328 but u-boot fails 
```
ModuleNotFoundError: No module named 'distutils.core'
scripts/dtc/pylibfdt/Makefile:27: recipe for target 'scripts/dtc/pylibfdt/_libfdt.so' failed
```
do you think we will need **host** python3-distutils python3-devel and swig or this can be patched in u-boot ?

diizzyy commented:
Regarding the atf files, we could alternatively just use FreeBSD's packages instead unless someone knows a better source? These are going to get bumped to 2.3 as soon as it's released 

https://freebsd.pkgs.org/13/freebsd-amd64/atf-rk3399-v2.1.txz.html
https://freebsd.pkgs.org/13/freebsd-amd64/atf-rk3328-v2.1.txz.html

lucize commented:
also I get (remember for rk3328)
```
U-Boot TPL 2020.01 (Apr 15 2020 - 10:47:43)
data training error
col error
data training error
LPDDR3, 800MHz
BW=16 Col=12 Bk=8 CS0 Row=16 CS=1 Die BW=8 Size=4096MB
```

lucize commented:
@diizzyy I think is a good call

Manawyrm commented:
@lucize 
> Regarding the atf files, we could alternatively just use FreeBSD's packages instead unless someone knows a better source? These are going to get bumped to 2.3 as soon as it's released

That looks quite good actually, but 2.1 is really too old, it wouldn't work well like that. 
Do you happen to know if FreeBSD has a beta or nightly repo with newer firmware?

lucize commented:
@Manawyrm  I don't know about nighlty 
so with self compiled I have almost the same, I'm missing something ? maybe a boot argument somewhere ?

```
U-Boot TPL 2020.01 (Apr 15 2020 - 10:47:43)
LPDDR3, 800MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS=1 Die BW=16 Size=1024MB
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.01 (Apr 15 2020 - 10:47:43 +0000)
Trying to boot from MMC1
NOTICE:  BL31: v2.2():reboot-12985-g508462a399
NOTICE:  BL31: Built : 10:47:43, Apr 15 2020
NOTICE:  BL31:Rockchip release version: v1.2


U-Boot 2020.01 (Apr 15 2020 - 10:47:43 +0000)

Model: Pine64 Rock64
DRAM:  1022 MiB
PMIC:  RK8050 (on=0x40, off=0x00)
MMC:   rksdmmc@ff500000: 1, rksdmmc@ff520000: 0
Loading Environment from MMC... *** Warning - bad CRC, using default environment

In:    serial@ff130000
Out:   serial@ff130000
Err:   serial@ff130000
Model: Pine64 Rock64
Net:   eth0: ethernet@ff540000
Hit any key to stop autoboot:  0
Card did not respond to voltage select!
switch to partitions #0, OK
mmc1 is current device
Scanning mmc 1:1...
Found U-Boot script /boot.scr
366 bytes read in 4 ms (88.9 KiB/s)
## Executing script at 00500000
34692 bytes read in 8 ms (4.1 MiB/s)
12931080 bytes read in 1124 ms (11 MiB/s)
## Flattened Device Tree blob at 01f00000
   Booting using the fdt blob at 0x1f00000
   Loading Device Tree to 000000003df1b000, end 000000003df26783 ... OK

Starting kernel ...

```

Manawyrm commented:
@lucize Should we maybe move this discussion to another PR? I'd suggest you open a WIP PR and we can work there without bloating this PR.

diizzyy commented:
@Manawyrm 
As I said, it's going to be updated shortly once 2.3 is out which I guess should be in 2 weeks or so.

lucize commented:
@Manawyrm I'll open another PR, for now it seems that setenv bootargs has some problem, without it the kernel starts but is panicing in finding rootfs (as expected), I'll test some more

adrianschmutzler commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,26 @@
+include $(INCLUDE_DIR)/target.mk
+
+DEFAULT_PACKAGES += \
+	uboot-envtools partx-utils e2fsprogs
```
If it's limited to those few, you may put everything in one line ...

Manawyrm replied:
Yup, done.

adrianschmutzler commented on target/linux/rockchip/image/mmc.bootscript:
```diff
@@ -0,0 +1,9 @@
+load mmc ${devnum}:1 ${kernel_addr_r} kernel.img
+
+booti ${kernel_addr_r} - ${fdt_addr_r}
+
```
Remove empty line at the end :-)

Manawyrm replied:
heh :)

Manawyrm commented:
@ynezz @adrianschmutzler 

@TobleMiner and I have built a reproducible and generalized build method for the ATFs and have setup a GitHub repository with CI for it: 
https://github.com/atf-builds/atf

Please take a look at the setup and the new arm-trusted-firmware package for OpenWRT. 
This could also be extended for all the other ATF packages (mvebu) pretty easily.

ynezz commented:
>  reproducible and generalized build method for the ATFs and have setup a GitHub repository with CI for it

Thats really nice, thanks! I've just build tested it (had to revert d27623b542548b765ddd46b046169006a3b5f66f), but otherwise looks good to me.

`Acked-by: Petr Štetiar <ynezz@true.cz>`

ynezz commented:
BTW is is possible to source this device somewhere in EU? I couldn't find that info.

tmn505 commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,25 @@
+ARCH:=arm
+BOARD:=rockchip
+BOARDNAME:=Rockchip RK33xx
+FEATURES:=ext4 audio usb usbgadget display gpio fpu rootfs-part boot-part
```
Please add squashfs for images with f2fs overlay.

Manawyrm replied:
Done.

tmn505 commented on target/linux/rockchip/Makefile:
```diff
@@ -0,0 +1,25 @@
+BOARD:=rockchip
+BOARDNAME:=Rockchip RK33xx
+FEATURES:=ext4 audio usb usbgadget display gpio fpu rootfs-part boot-part
+SUBTARGETS:=rk3399
```
Why creating subtarget when there is only single device in this target? Do You have upcoming additions of devices in queue?

diizzyy replied:
RK3328 for instance?

tmn505 replied:
Do You have it in Your queue? Also dropping the `CPU_TYPE` would allow to support rk3328 under the same target and this device is not space constrained to justify fragmentation. It works well for layerscape.

lucize replied:
I have a "working" rk3328, the ethernet is a problem atm

tmn505 replied:
Good, still that doesn't justify fragmentation, with current state of this PR You could use rk3399 subtaget and simply add necessary drivers and be done. But that's not rk3288 which would justify **subtargets**. For now **single** target is enough.

Manawyrm replied:
Yes, we're planning on adding more hardware in the near future. I've spoken to Pine64 (we're also part of the Manjaro project) and they're sending some more Rockchip hardware my way. 
One of these platforms needs it's own subtarget and has space constraints (because the image needs to fit on an SPI flash).

blocktrron replied:
@tmn505 Currently every target present in OpenWrt has at least one subtarget (e.g. `ipq806x-generic` or `ipq40xx-generic`).

So before we are doing stuff like `rockchip-generic` I'm in favor of the current state.

tmn505 replied:
@Manawyrm that's nice, it's good to see vendors willing to give hardware for open contributions.
That is why I also suggested tuning the kernel configuration. Also if need will arise, You can move drivers not needed to boot to kernel modules which will occupy rootfs or even extroot. If U-Boot is not constrained in regard of kernel size, kernel can be as big as 15MB if the flash size is 16MB. Creating additional subtarget should be last option.

@blocktrron Ok, then the name should change. With current state, this subtarget would also support `rk3328` but the name suggests only `rk3399` which is misleading. Going further, if `CPU_TYPE` would be removed, in the same subtarget we could also add `rk1808` devices, then again the name would be misleading. We could change that as the new devices would be added but this would leave dangling directories for targets no longer updated, occupying storage on download server. It's best to chose right name at the beginning. I would opt for `armv8` or `64` or both combined, it the need would arise then something like `armv8_tiny` could be created.

ynezz replied:
@tmn505 has a good point, looking around, what about `cortexa53` subtarget as seen in sunxi, mvebu already? Are those bad examples?

ynezz replied:
If this subtarget could support rk1808 as well, which is `cortex-a35`, then the proposed `cortexa53` doesnt make much sense anymore, so probably the remaining ones as already suggested `64` or `armv8_64`.

Manawyrm replied:
I went with cortexa53. I've spoken to @TobleMiner and some Pine64 folks and there's going to be some future additional hardware support that will require different subtargets.

ynezz replied:
@Manawyrm That mine initial idea with `cortex-a53` wasn't that good as @tmn505 explained, there is RK1808 which could be supported with this subtarget as well as it is dual-core ARM `cortex-a35`, so my [previous comment](https://github.com/openwrt/openwrt/pull/2917#discussion_r410300701) still stands.

There seems to be [TB-96AIoT](https://www.seeedstudio.com/BeiQi-RK1808-AIoT-96Boards-Compute-SoM-p-4074.html) already available from Seeed and [SoEdge-RK1808 SoM](https://www.cnx-software.com/2019/08/05/pine64-soedge-rk1808-module-delivers-3-0-tops-via-rockchip-rk1808-soc) from pine64.

> there's going to be some future additional hardware support that will require different subtargets

Ok, then you need to provide more details then just someone told me so.

Manawyrm replied:
@ynezz Hm, ok. Let's just make it generic then. 
How do we want to name the subtarget? 

armv8_64b ? (identical to layerscape?)

ynezz replied:
Naming is hard :) I would just go with `armv8`, it's probably unlikely, that Rockchip is going to use some older 32-bit ARMv8 core.

TobleMiner replied:
`armv8` sounds good to be, too

tmn505 commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,51 @@
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/image.mk
+
+FAT32_BLOCK_SIZE=1024
```
Does the bootflow require the first partition to be FAT formatted? If not, the preference would be using `scripts/gen_image_generic.sh`, instead of introducing another device specific script. One exception would probably be EFI boot, if someone would like to add support for such images.

Manawyrm replied:
No, we don't need FAT, we could also use ext4 or something else.
But what we do need is the SoC-specific script that dd's the bootloader to magic locations on the MMC image. So I fear we can't get rid of the script yet, sorry.

tmn505 replied:
An example here: https://github.com/openwrt/openwrt/blob/0495324b9bd0b6a976ea731a1b4b77b26df32ee6/target/linux/tegra/image/Makefile#L33
You could add it as separate command `uboot-install`.

Manawyrm replied:
Switched to EXT4.

tmn505 commented on target/linux/rockchip/rk3399/target.mk:
```diff
@@ -0,0 +1,9 @@
+ARCH:=aarch64
+SUBTARGET:=rk3399
+BOARDNAME:=RK3399 boards (64 bit)
+CPU_TYPE:=cortex-a53
```
Drop the `CPU_TYPE`, this will allow for multiple ARM64 core types to be supported from single target now and probably in the future. This mistake has already been made for mvebu which puts additional strain on buildbots.

diizzyy replied:
Why would we drop CPU_TYPE?

tmn505 replied:
Did You read second sentence?

ynezz replied:
I don't follow here as well:

```
target/linux/armvirt/64/target.mk  ARCH:=aarch64
target/linux/bcm27xx/bcm2710/target.mk CPU_TYPE:=cortex-a53 ARCH:=aarch64
target/linux/bcm27xx/bcm2711/target.mk CPU_TYPE:=cortex-a72 ARCH:=aarch64
target/linux/ipq807x/Makefile CPU_TYPE:=cortex-a53 ARCH:=aarch64
target/linux/layerscape/armv8_64b/target.mk  ARCH:=aarch64
target/linux/mediatek/mt7622/target.mk CPU_TYPE:=cortex-a53 ARCH:=aarch64
target/linux/mvebu/cortexa53/target.mk CPU_TYPE:=cortex-a53 ARCH:=aarch64
target/linux/mvebu/cortexa72/target.mk CPU_TYPE:=cortex-a72 ARCH:=aarch64
target/linux/octeontx/Makefile  ARCH:=aarch64
target/linux/rockchip/rk3399/target.mk CPU_TYPE:=cortex-a53 ARCH:=aarch64
target/linux/sunxi/cortexa53/target.mk CPU_TYPE:=cortex-a53 ARCH:=aarch64
```

Manawyrm replied:
What does dropping CPU_TYPE do? Does it use the lowest instruction set supported by all aarch64 CPUs? Or something else?

tmn505 replied:
@ynezz Look at my answer to blocktrron, it's always best to support wide spectrum of SoCs with single target, which would have benefit of less strain on buldbots and download server.

@Manawyrm Yes.

ynezz replied:
> which would have benefit of less strain on buldbots and download server

@tmn505 But as you can see from my list above, it makes no sense to me. We've already plenty of aarch64 targets which use CPU_TYPE to enable GCC performance tuning already. So what would we gain by removing CPU_TYPE in this target? Probably just attract hordes of benchmark freaks, which would run in the circles and screaming something about intentional crippled performance of their CPUs :-)

tmn505 replied:
@ynezz Yes, we do have them. For example rk1808 is Cortex A35 would need additional subtarget, where toolchain and packages would need to be built from ground up. With dropping CPU_TYPE you could potentially support it under the same target. You can't please everyone, they do complain but at the end of the day part of them still use Debian or other main distro, which doesn't have that fine tuning. If the hordes want the fine tuning, there's nothing against adding optimizations to local build. BTW. I'm more afraid of jow- than the hordes :-)

ynezz replied:
> rk1808 is Cortex A35
> With dropping CPU_TYPE you could potentially support it under the same target.

Ok, now this makes sense.

Manawyrm replied:
I have dropped the CPU_TYPE.

diizzyy replied:
@tmn505 @ynezz 
If that's the case we can add it to new (unique) subtargets, there's no point in doing to for already existing ones and if we're going to use Debian as an example, they don't as far as I kno enforce -Os either if you care about performance which many seems to do...

tmn505 reviewed (changes requested):
Hi Tobias.
Nice that ATF build has been sorted out. Aside from my comments I would suggest fine tuning kernel config.

diizzyy commented:
@ynezz 
https://sha256systems.eu/epages/5b7c8826-3528-4d6f-bd2d-c024fcdbd435.sf/de_DE/?ObjectPath=/Shops/5b7c8826-3528-4d6f-bd2d-c024fcdbd435/Products/RPro64-4GB
https://botland.com.pl/en/pine64-rock64/12991-pine64-rockpro64-rockchip-rk3399-cortex-a72-a53-4gb-ram.html?search_query=rockpro64&results=16

diizzyy commented:
As much it fixes our ATF firmware issue right now I'm not too fond of our own fork as this will quite easily get unmaintained.

lucize commented:
> I would suggest fine tuning kernel config.

maybe enable hw crypto in kernel ?
```

Embedded dual-channel encryption and decryption engine
 Support AES 128/192/256 bits key mode, ECB/CBC/CTR/XTS chain mode,
Slave/FIFO mode
 Support DES/3DES(ECB and CBC chain mode), 3DES(EDE/EEE key mode),
Slave/FIFO mode
 Support SHA1/SHA256/MD5(with hardware padding) HASH function, FIFO mode
only
 Support 160 bit Pseudo Random Number Generator (PRNG)
 Support 256 bit True Random Number Generator (TRNG)
 Support PKA 512/1024/2048 bit Exp Modulator
```

diizzyy commented:
@lucize 
Can you verify that it actually works and provides a benefit?

lucize commented:
I'll test, I think  that even if it is offloading the cpu is a good thing

ynezz commented:
> As much it fixes our ATF firmware issue right now I'm not too fond of our own fork as this will quite easily get unmaintained

What do you suggest?

I'm probably missing something here, but there is no fork happening, it's just reproducible [build](https://github.com/atf-builds/atf/blob/master/build.sh) machinery, which produces binary artifacts which could be reused by other projects/people building own bootloaders but not willing to build ATFs themselves, CI pipelines etc. The key feature here is, that the binaries are reproducible.

If this [atf-builds/atf](https://github.com/atf-builds/atf) project gets unmaintained in the future for whatever reason, one could ask for membership in that project to continue the maintenance or this could be forked etc. Not a big deal I would say.

diizzyy commented:
@ynezz 
Which is why I suggested that grabbing it from FreeBSD (or another distro if available) instead of reinventing the wheel again just because.

Manawyrm commented:
@diizzyy @ynezz 
> As much it fixes our ATF firmware issue right now I'm not too fond of our own fork as this will quite easily get unmaintained.

@TobleMiner and me are both on GitHub for almost 10 years now and we're not going anywhere soon :) 
If anything catastrophic (car crash, etc.) happens, it's only a matter of forking the repo or downloading the Repo and reuploading it somewhere else (for example in the OpenWRT account). 
The build scripts will automatically start building new binaries in the new repository on GitHub. 

OpenWRT still has the absolute control, because it has the URLs and SHA256SUMs in the package source. 
So yeah, if anything happens, just change the URL to a new repository.

Manawyrm commented:
> @ynezz
> Which is why I suggested that grabbing it from FreeBSD (or another distro if available) instead of reinventing the wheel again just because.

FreeBSD is nice, but at the moment they only have very old software and I don't want to wait for them to release new binaries. 
They also don't have ATFs for RK3328, RK1808 or any of the Marvell CPUs. 

I'm planning on adding support for a handful of more Rockchip SoCs into OpenWRT soon. 
This solution also applies to the mvebu repo (well not yet, but soon).

lucize commented:
@Manawyrm does your board have working usb3 ? I've added CONFIG_PHY_ROCKCHIP_INNO_USB3 and the following patch https://patchwork.kernel.org/cover/11216245/ but it also seem to need some dts patching.
after adding the usb3 patch and enable 
```
[    0.400632] rockchip-u3phy ff470000.usb3-phy: vbus_drv is not assigned
[    0.403374] rockchip-u3phy ff470000.usb3-phy: Rockchip u3phy initialized successfully
```
maybe still some usb3hub needed in dts ?
but also I have
`[    0.429570] spi-nor spi0.0: gd25q128 (16384 Kbytes)`
does it means that I have a spi flash on my board and I could flash this to the spi chip?

lucize commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -0,0 +1,55 @@
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+PKG_VERSION:=2020.01
```
2020.04 is now available

ynezz replied:
Indeed, but it is out of scope of this PR and AFAIK there is no such rule, that the PR needs to contain latest&greatest versions of all packages, although it would be certainly nice.

lucize replied:
yes, but with 2020.04, there is no need for patching so rk3328 boots fine

ynezz replied:
This PR is about RockPro64.

lucize replied:
Yes, but then I'll have to modify this file with the upcoming PR for rk3328, I'm just waiting for this PR to come to a conclusion

Manawyrm replied:
I will try a modern u-boot release and then update this version number.

lucize replied:
what about python3-dev and swig, do you already have them installed on your host, I can't build u-boot on my machine without them, or is a normal dependency for openwrt ? does the buildbot have them installed ?

tmn505 replied:
I don't know if python headers are installed, but swig is definitely not. Sunxi target has a patch for that and tegra sed's out BINMAN from Kconfig. I don't know if that's feasible with rockchip.

Manawyrm replied:
Very cool! Thank you very much for that info about BINMAN. 
I've added that to rockchip and tested it. It seems to work. 

@lucize As your system didn't have python3-dev and swig, could you try again if it builds for you now?

lucize replied:
so I added the
```
define Build/Configure
        $(SED) '/select BINMAN/d' -i $(PKG_BUILD_DIR)/arch/arm/mach-rockchip/Kconfig
        $(call Build/Configure/U-Boot)

```
but

```
Traceback (most recent call last):
  File "scripts/dtc/pylibfdt/setup.py", line 21, in <module>
    from distutils.core import setup, Extension
ModuleNotFoundError: No module named 'distutils.core'
make[3]: Leaving directory '/home/build/source/feeds/packages/utils/pciutils'
time: package/feeds/packages/pciutils/compile#0.26#0.21#0.63
scripts/dtc/pylibfdt/Makefile:27: recipe for target 'scripts/dtc/pylibfdt/_libfdt.so' failed
make[7]: *** [scripts/dtc/pylibfdt/_libfdt.so] Error 1
scripts/Makefile.build:432: recipe for target 'scripts/dtc/pylibfdt' failed

```

lucize replied:
I also tried to remove select SPL_OF_LIBFDT and select TPL_OF_LIBFDT I guess I'll try to research a bit more on this

Manawyrm replied:
@lucize I think we have a solution. Could you pull the git, clean and try to build again?

lucize replied:
I think FDT needs to be disabled everywhere at least is not needed in the boot script

```
Traceback (most recent call last):
  File "./tools/dtoc/dtoc", line 44, in <module>
    import dtb_platdata
  File "/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rock64-rk3328/u-boot-2020.04/tools/dtoc/dtb_platdata.py", line 18, in <module>
    import fdt
  File "/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rock64-rk3328/u-boot-2020.04/tools/dtoc/fdt.py", line 12, in <module>
    import libfdt
ModuleNotFoundError: No module named 'libfdt'
Traceback (most recent call last):
  File "./tools/dtoc/dtoc", line 44, in <module>
    import dtb_platdata
  File "/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rock64-rk3328/u-boot-2020.04/tools/dtoc/dtb_platdata.py", line 18, in <module>
    import fdt
  File "/home/build/source/build_dir/target-aarch64_cortex-a53_musl/u-boot-rock64-rk3328/u-boot-2020.04/tools/dtoc/fdt.py", line 12, in <module>
    import libfdt
ModuleNotFoundError: No module named 'libfdt'
scripts/Makefile.spl:324: recipe for target 'tpl/dts/dt-platdata.c' failed
make[5]: *** [tpl/dts/dt-platdata.c] Error 1
make[5]: *** Waiting for unfinished jobs....
  CC      spl/disk/part_dos.o
scripts/Makefile.spl:321: recipe for target 'include/generated/dt-structs-gen.h' failed
make[5]: *** [include/generated/dt-structs-gen.h] Error 1
Makefile:1876: recipe for target 'tpl/u-boot-tpl.bin' failed
make[4]: *** [tpl/u-boot-tpl.bin] Error 2
make[4]: *** Waiting for unfinished jobs....

```

Manawyrm replied:
Sorry, I can't reproduce this with the latest debian version.
I've installed a fresh Debian 10, installed git, gawk, ncurses and build-essential and it just worked.

Are you sure you're using the latest git?

lucize replied:
I used 
```
define Build/Configure
        $(SED) s/subdir.*pylibfdt//g $(PKG_BUILD_DIR)/scripts/dtc/Makefile
```
but jut want to let you know that rk3399 is working ok, adding 
```
  BUILD_DEVICES:= \
    rock64
```
will need the python3-dev and swig so this can go in and just see about rk3328 later

ynezz reviewed (changes requested):
@tmn505 has provided additional feedback which makes sense to me as well, please address his remarks as well, while at it adjust the commit message as well `U-boot and ATF are build from the mainline trees and integrated into the images.` as this no longer applies anymore.

Manawyrm commented:
@lucize 
> does it means that I have a spi flash on my board and I could flash this to the spi chip?

Yes! Both RockPro64 and Rock64 have 16 MByte (128 MBit) SPI NOR flash and OpenWRT can be booted from them. 
I will implement this very soon as I have a new platform that will require this :) 
That will be implemented in a seperate PR.

Manawyrm commented:
@ynezz Hi! Could you check again if everything is OK now?

Manawyrm commented:
The kernel compiles to a 12.5 MiB file, which can be compressed to 3.9 MiB with LZMA, which is very reasonable for the RK3399 SoC. 

U-boot is now 2020.04, u-boot build works without python. 
squashfs support is enabled, tested and works. 
 
Everything OK now?

(@ynezz please check again :smile: )

diizzyy commented:
@Manawyrm 
RK3328 is there
https://github.com/openwrt/openwrt/pull/2917#issuecomment-614022072 however Marvell isn't so do as you wish.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_CAVIUM_TX2_ERRATUM_219=y
+CONFIG_CC_DISABLE_WARN_MAYBE_UNINITIALIZED=y
+CONFIG_CC_HAS_KASAN_GENERIC=y
+# CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE is not set
+CONFIG_CC_OPTIMIZE_FOR_SIZE=y
```
Is this intentional decision? Generic config sets performance, check target/linux/generic.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_CC_HAS_KASAN_GENERIC=y
+# CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE is not set
+CONFIG_CC_OPTIMIZE_FOR_SIZE=y
+CONFIG_CDROM=y
```
Disable.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_BRCMSTB_GISB_ARB=y
+CONFIG_BSD_PROCESS_ACCT=y
+CONFIG_BSD_PROCESS_ACCT_V3=y
+CONFIG_BTRFS_FS=y
+# CONFIG_BTRFS_FS_CHECK_INTEGRITY is not set
+CONFIG_BTRFS_FS_POSIX_ACL=y
```
Disable btrfs support. You can install it with opkg and it's not required to boot the device.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_EXT4_FS_POSIX_ACL=y
+CONFIG_EXTCON=y
+CONFIG_FANOTIFY=y
+CONFIG_FAT_FS=y
```
Disable, not needed anymore, since boot fs is ext4.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_EMAC_ROCKCHIP=y
+CONFIG_ENABLE_MUST_CHECK=y
+CONFIG_ENERGY_MODEL=y
+CONFIG_EXT3_FS=y
+# CONFIG_EXT3_FS_POSIX_ACL is not set
+# CONFIG_EXT3_FS_SECURITY is not set
```
Disable ext3, ext4 driver can handle it.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_FS_IOMAP=y
+CONFIG_FS_MBCACHE=y
+CONFIG_FS_POSIX_ACL=y
+CONFIG_FUJITSU_ERRATUM_010001=y
```
Disable.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_FS_POSIX_ACL=y
+CONFIG_FUJITSU_ERRATUM_010001=y
+CONFIG_FW_LOADER_PAGED_BUF=y
+CONFIG_GCC_PLUGINS=y
+# CONFIG_GCC_PLUGIN_CYC_COMPLEXITY is not set
+# CONFIG_GCC_PLUGIN_LATENT_ENTROPY is not set
+# CONFIG_GCC_PLUGIN_RANDSTRUCT is not set
+# CONFIG_GCC_PLUGIN_STACKLEAK is not set
+# CONFIG_GCC_PLUGIN_STRUCTLEAK_BYREF is not set
+# CONFIG_GCC_PLUGIN_STRUCTLEAK_BYREF_ALL is not set
+# CONFIG_GCC_PLUGIN_STRUCTLEAK_USER is not set
```
Remove, let it be set on the fly.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_HAVE_UID16=y
+CONFIG_HAVE_VIRT_CPU_ACCOUNTING_GEN=y
+CONFIG_HID=y
+CONFIG_HID_A4TECH=y
+CONFIG_HID_BELKIN=y
+CONFIG_HID_CHERRY=y
+CONFIG_HID_CHICONY=y
+CONFIG_HID_CYPRESS=y
+CONFIG_HID_EZKEY=y
```
You can drop these, since most devices will be handled by generic driver.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_HID_CYPRESS=y
+CONFIG_HID_EZKEY=y
+CONFIG_HID_GENERIC=y
+CONFIG_HID_ITE=y
+CONFIG_HID_KENSINGTON=y
+CONFIG_HID_LOGITECH=y
+CONFIG_HID_MICROSOFT=y
+CONFIG_HID_MONTEREY=y
+CONFIG_HID_REDRAGON=y
```
As above comment.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_HID_MICROSOFT=y
+CONFIG_HID_MONTEREY=y
+CONFIG_HID_REDRAGON=y
+CONFIG_HISILICON_ERRATUM_161600802=y
```
Disable.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_MFD_SYSCON=y
+# CONFIG_MFD_VEXPRESS_SYSREG is not set
+CONFIG_MIGRATION=y
+CONFIG_MISC_RTSX=y
+CONFIG_MISC_RTSX_USB=y
```
You can probably remove these.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_NEED_SG_DMA_LENGTH=y
+CONFIG_NET_FLOW_LIMIT=y
+CONFIG_NLS=y
+CONFIG_NLS_CODEPAGE_437=y
+CONFIG_NLS_ISO8859_1=y
```
Since FAT is out You can drop NLS support.

Manawyrm replied:
Selected by [y]:
- F2FS_FS [=y] && BLOCK [=y]  

F2FS reenables it. Unfortunate, but ok.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_PWM=y
+CONFIG_PWM_ROCKCHIP=y
+CONFIG_PWM_SYSFS=y
+CONFIG_QCOM_FALKOR_ERRATUM_1003=y
+CONFIG_QCOM_FALKOR_ERRATUM_1009=y
+CONFIG_QCOM_FALKOR_ERRATUM_E1041=y
+CONFIG_QCOM_QDF2400_ERRATUM_0065=y
```
Disable.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_XHCI_PLATFORM=y
+# CONFIG_USERIO is not set
+CONFIG_VEXPRESS_CONFIG=y
+CONFIG_VEXPRESS_SYSCFG=y
```
These can probably be removed.

tmn505 commented on target/linux/rockchip/cortexa53/config-5.4:
```diff
@@ -0,0 +1,817 @@
+# CONFIG_USERIO is not set
+CONFIG_VEXPRESS_CONFIG=y
+CONFIG_VEXPRESS_SYSCFG=y
+CONFIG_VFAT_FS=y
```
Disable, not used to boot anymore.

tmn505 commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,65 @@
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/image.mk
+
+FAT32_BLOCK_SIZE=1024
+FAT32_BLOCKS=$(shell echo $$(($(CONFIG_TARGET_KERNEL_PARTSIZE)*1024*1024/$(FAT32_BLOCK_SIZE))))
```
Remove, not used anymore.

tmn505 commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,65 @@
+FAT32_BLOCK_SIZE=1024
+FAT32_BLOCKS=$(shell echo $$(($(CONFIG_TARGET_KERNEL_PARTSIZE)*1024*1024/$(FAT32_BLOCK_SIZE))))
+
+DEVICE_VARS += UBOOT_DEVICE_NAME SOC
```
Remove SOC it's already set in include/image.mk.

tmn505 commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,65 @@
+endef
+
+### Devices ###
+define Device/Default
+  KERNEL := kernel-bin
+  IMAGES := sysupgrade.img.gz
+  SUPPORTED_DEVICES := $(subst _,$(comma),$(1))
+  DEVICE_DTS = rockchip/$$(SOC)-$(lastword $(subst _, ,$(1)))
+endef
```
Add here `PROFILES := Default`, there is some bug in OpenWrt buildroot which is not yet tracked, and it's causing only the first device definition to be evaluated.

tmn505 reviewed (changes requested):
The kernel config comment are more like suggestions to slim it down, not requests. You don't have to address them, I just posted them since I won't be looking at this target anymore, unless I'll buy device with this SoC.
But do please address the other comments.

Manawyrm commented:
@tmn505 Thank you very much for the suggestions, they were very helpful :+1: 
I have now addressed them and pushed a new version of my PR.

diizzyy commented:
Please don't use armv8 as it will disable optimizations, we already have cortex-a53 in tree so there's no additional load by adding it.

Manawyrm commented:
@diizzyy Please discuss this with @tmn505 . 
There was a big discussion here: 
https://github.com/openwrt/openwrt/pull/2917#discussion_r409703705
This is the 3rd time I've changed the subtarget now, it's getting a bit annoying by now.

lucize commented on target/linux/rockchip/image/mmc.bootscript:
```diff
@@ -0,0 +1,8 @@
+part uuid mmc ${devnum}:2 uuid
+
+setenv bootargs "console=ttyS2,1500000 console=tty1 earlycon=uart8250,mmio32,0xff1a0000 root=PARTUUID=${uuid} rw rootwait"
```
can we make 0xff1a0000 from earlycon as a variable and pass it by Makefile ? for rk3328 we need 0xff130000 so we can have only one mmc script

Manawyrm replied:
Yes, but please do that in your own PR.

lucize commented on target/linux/rockchip/image/Makefile:
```diff
@@ -0,0 +1,63 @@
+endef
+
+define Build/boot-script
+	# Make an U-boot image and copy it to the FAT32 partition
```
there is no more FAT32

Manawyrm replied:
Yup, done, thanks.

adrianschmutzler commented on target/linux/rockchip/armv8/base-files/lib/upgrade/platform.sh:
```diff
@@ -0,0 +1,86 @@
+
+	#copy partition uuid
+	echo "Writing new UUID to /dev/$diskdev..."
+	get_image "$@" | dd of="/dev/$diskdev" bs=1 skip=440 count=4 seek=440 conv=fsync
```
Maybe bs=4 skip=110 seek=110 count=1

Manawyrm replied:
bcm27xx, sunxi, tegra, apm821xx, omap, mvebu and x86 have the same file. 
Personally I find offset 440 to be more readable, are the 3 additional write()-syscalls so bad? 

I could open a new PR to change this in all targets, if you'd like?

adrianschmutzler replied:
Don't think it will have any measurable effect, I just wanted to mention it in case you aren't aware since the option to use bs instead of count is frequently overlooked (and we had cases where it was relevant).

So, if you chose it by intention, I'm perfectly fine (and the thing is merged anyway). That's why I chose the "Maybe" for my comment :-)

Manawyrm replied:
OK, i don't think it'll have any relevant impact, only case where it would make sense would be SPI NOR flash (you'd erase the same flash block 4 times instead of once), but with SD and eMMC storage the block layer will handle this properly.

ynezz commented:
Merged into my staging tree, thanks!

ynezz commented:
> Please don't use armv8 as it will disable optimizations, we already have cortex-a53 in tree so there's no additional load by adding it.

You've probably missed this part:

> rk1808 is Cortex A35. With dropping CPU_TYPE you could potentially support it under the same target.

There is no such `cortex-a35` CPU type in the tree, thats why `armv8` seems just good enough to me at this moment. It's good balance between **general** usability and current project resources.

BTW this is my final comment on this PR, should you need to discuss this more, please bring this topic on the mailing list.

Manawyrm commented:
Thanks! :smile:

diizzyy commented:
Unless RK1808 is going to use the same kernel config(?) it still needs a subtarget which can be set to armv8? Also, shouldn't it say armv8-a?

Manawyrm commented:
> Also, shouldn't it say armv8-a?
This was discussed here:
https://github.com/openwrt/openwrt/pull/2917#discussion_r410658686
(and since Rockchip don't make any ARMv8 32bit cores, it's unambiguous.)

diizzyy commented:
That reason I'm asking is because I can't find if armv8 is a valid option or if it just happends to work by accident? https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/AArch64-Options.html

Manawyrm commented:
The ARCH-variable in https://github.com/Manawyrm/openwrt/blob/a8ac4868d58a33c19ede3d6241afba28c46a75b9/target/linux/rockchip/armv8/target.mk#L1 causes it to be compiled for the right architecture, AFAIK the subtarget-name is just a name and has no actual relevance.

diizzyy commented:
Hmm.. I'll try to have a look at GCC code however as far as I can tell by the documentation seems that it defines crypto instructions as an example as armv8-a extensions

rdslw commented:
@Manawyrm how far from this PR would be adding support for kobol's Helios63, which is also RK3399 device?

karthirockz commented:
@Manawyrm 

add support for rk3288

http://en.t-firefly.com/product/industry/aio_3288c.html?theme=pc
